ambry-api/src/main/java/com/github/ambry/server/StatsSnapshot.java

```diff
@@ -151,15 +150,11 @@ public void removeZeroValueSnapshots() {
     // if we are here, then value is not 0, we can examine all the StatsSnapshots in the subMap and remove the 0-value
     // StatsSnapshot.
     if (subMap != null) {
-      List<String> keysToRemove = new ArrayList<>();
-      for (Map.Entry<String, StatsSnapshot> ent : subMap.entrySet()) {
-        if (ent.getValue().getValue() == 0) {
-          keysToRemove.add(ent.getKey());
-        } else {
-          ent.getValue().removeZeroValueSnapshots();
-        }
-      }
-      keysToRemove.forEach(k -> subMap.remove(k));
+      subMap = subMap.entrySet()
+          .stream()
+          .filter(ent -> ent.getValue().getValue() != 0)
```

Comment:
Another idea to consider (for future PR) is to exclude containers from stats report if value is below a configurable threshold (like 100MB) rather than only zero.  That way, for accounts with many containers, only those with storage above the threshold will be counted (and ones with very little data yet will be excluded).

Reply:
no, sorry about this one. The reason to remove the zero value snapshots is because in bucket reporting, lots of containers might have 0 values and still be included in the reports. Before bucket reporting, if a container has 0 byte usage, then it will not present in the map. But in bucket report, since we keep the previous container usage values and apply the delta values in each bucket, there are chances that containers previously have non-0 values but now have 0 value would be included. This is to fix that. For instance, assume each bucket span is 10 minutes and container A has only one blob(4mb) that would expire in 10 minutes. So if we fetch usages now, we will see container A's value is 4m. Since this blob will be expired in next bucket, when fetching usages after 10 minutes, we will apply the delta in next bucket (-4m) and then get a 0 value. In this case, we don't need to include this container.

Your idea seems quite reasonable. We are giving each container a minimal 1GB as its quota. If a container' usage is less than some threshold,(maybe 500MB, we can do some calculations) we don't have to include it. This will work just fine for storage quota.